bson/src/main/org/bson/codecs/pojo/ConventionAnnotationImpl.java

```diff
@@ -111,49 +112,104 @@ private void processPropertyAnnotations(final ClassModelBuilder<?> classModelBui
+                    for (Annotation annotation : method.getDeclaredAnnotations()) {
+                        if (annotation.annotationType().equals(BsonCreator.class)) {
+                            if (creatorExecutable != null) {
+                                throw new CodecConfigurationException("Found multiple constructors / methods annotated with @BsonCreator");
+                            } else if (!bsonCreatorClass.isAssignableFrom(method.getReturnType())) {
+                                throw new CodecConfigurationException(
+                                        format("Invalid method annotated with @BsonCreator. Returns '%s', expected %s",
+                                                method.getReturnType(), bsonCreatorClass));
+                            }
+                            creatorExecutable = new CreatorExecutable<T>(clazz, method);
+                            foundStaticBsonCreatorMethod = true;
+                            break;
                         }
-                        creatorExecutable = new CreatorExecutable<T>(clazz, method);
-                        break;
                     }
                 }
             }
+
+            bsonCreatorClass = bsonCreatorClass.getSuperclass();
         }
 
         if (creatorExecutable != null) {
             List<BsonProperty> properties = creatorExecutable.getProperties();
             List<Class<?>> parameterTypes = creatorExecutable.getParameterTypes();
+            List<Type> parameterGenericTypes = creatorExecutable.getParameterGenericTypes();
+
             if (properties.size() != parameterTypes.size()) {
                 throw creatorExecutable.getError(clazz, "All parameters must be annotated with a @BsonProperty in %s");
             }
             for (int i = 0; i < properties.size(); i++) {
-                BsonProperty bsonProperty = properties.get(i);
+                boolean isIdProperty = creatorExecutable.getIdPropertyIndex() != null && creatorExecutable.getIdPropertyIndex().equals(i);
                 Class<?> parameterType = parameterTypes.get(i);
-                PropertyModelBuilder<?> propertyModelBuilder = classModelBuilder.getProperty(bsonProperty.value());
-                if (propertyModelBuilder == null) {
-                    addCreatorPropertyToClassModelBuilder(classModelBuilder, bsonProperty.value(), parameterType);
-                } else if (!propertyModelBuilder.getTypeData().isAssignableFrom(parameterType)) {
+                Type genericType = parameterGenericTypes.get(i);
+                PropertyModelBuilder<?> propertyModelBuilder = null;
+
+                if (isIdProperty) {
+                    propertyModelBuilder = classModelBuilder.getProperty(classModelBuilder.getIdPropertyName());
+                } else {
+                    BsonProperty bsonProperty = properties.get(i);
+
+                    // Find the property using write name and falls back to read name
```

Comment:
We now prefer the write name first for properties. This removes the confusion about which of the 3 property names should be used with `@BsonCreator`.  Legacy property names are still supported if needs be.